File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    
    void find(vector<int> nums, int index, vector<int> ans, vector<vector<int>>& result){
        if(index>=nums.size()){
            result.push_back(ans);
            return;
        }
        
        //exclude
        find(nums,index+1,ans,result);
        
        //include
        ans.push_back(nums[index]);
        find(nums,index+1,ans,result);
    }
    
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int>ans;
        find(nums,0,ans,result);
        return result;
    }
};

File: 0078-subsets/0078-subsets.java
class Solution {
    
    public void solve(int idx, int[] nums, List<List<Integer>> ans, List<Integer> res){
        if(idx == nums.length){
            ans.add(new ArrayList<>(res));
            return;
        }
        
        res.add(nums[idx]);
        solve(idx+1,nums,ans,res);
        res.remove(res.size()-1);
        
        solve(idx+1,nums,ans,res);
    }
    
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> res = new ArrayList<>();
        
        solve(0,nums,ans,res);
        
        return ans;
    }
}